File: packages/polars/lib/src/list.dart
import 'wrapper/expr.dart';
import 'expr.dart';

class ListNamespace {
  final Expr expr;

  const ListNamespace(this.expr);

  Expr get len => expr.listLen;
  Expr get max => expr.listMax;
  Expr get min => expr.listMin;
  Expr get sum => expr.listSum;
  Expr get mean => expr.listMean;
  Expr get reverse => expr.listReverse;
  Expr get first => expr.listFirst;
  Expr get last => expr.listLast;
  Expr get argMin => expr.listArgMin;
  Expr get argMax => expr.listArgMax;

  Expr get(Object? index) => expr.listGet(index: index.expr);
  Expr join(Object? separator) => expr.listJoin(separator: separator.expr);
  Expr shift(Object? periods) => expr.listShift(periods: periods.expr);
  Expr slice(Object? offset, Object? length) =>
      expr.listSlice(offset: offset.expr, length: length.expr);
  Expr head([int n = 10]) => expr.listHead(n: n.expr);
  Expr tail([int n = 10]) => expr.listTail(n: n.expr);
  Expr contains(Object? item) => expr.listContains(other: item.expr);
  Expr unique({bool maintainOrder = false}) =>
      expr.listUnique(maintainOrder: maintainOrder);
}

File: polars-wrapper/src/wrapper/list.rs
use flutter_rust_bridge::frb;

use super::expr::Expr;

macro_rules! delegate_str {
    ($( $(#[$attribute:meta])* $fn:ident(self $(,)? $($param:ident : $(#[$conv:ident])? $ty:ty $(= $default:expr)? ),*) -> $output:ty; )*) => {paste::paste!{$(
        $(#[$attribute])*
        #[doc = concat!(" TODO: Docs for ", stringify!($fn))]
        pub fn [<list_ $fn>](&self, $($(#[frb(default = $default)])? $param : $ty),*) -> $output {
            <$output>::from(self.into_internal().list().$fn($($param $(.$conv())?),*))
        }
    )*}};
}

impl Expr {
    delegate_str! {
        #[frb(sync, getter)] len(self) -> Expr;
        #[frb(sync, getter)] max(self) -> Expr;
        #[frb(sync, getter)] min(self) -> Expr;
        #[frb(sync, getter)] sum(self) -> Expr;
        #[frb(sync, getter)] mean(self) -> Expr;
        // #[frb(sync, getter)] sort(self) -> Expr;
        #[frb(sync, getter)] reverse(self) -> Expr;
        #[frb(sync)] get(self, index: #[into] Expr) -> Expr;
        #[frb(sync, getter)] first(self) -> Expr;
        #[frb(sync, getter)] last(self) -> Expr;
        #[frb(sync)] join(self, separator: #[into] Expr) -> Expr;
        #[frb(sync, getter)] arg_min(self) -> Expr;
        #[frb(sync, getter)] arg_max(self) -> Expr;
        #[frb(sync)] shift(self, periods: #[into] Expr) -> Expr;
        #[frb(sync)] slice(self, offset: #[into] Expr, length: #[into] Expr) -> Expr;
        #[frb(sync)] head(self, n: #[into] Expr) -> Expr;
        #[frb(sync)] tail(self, n: #[into] Expr) -> Expr;
        #[frb(sync)] contains(self, other: Expr) -> Expr;
    }
    #[frb(sync)]
    pub fn list_unique(self, #[frb(default = false)] maintain_order: bool) -> Expr {
        if maintain_order {
            self.into_internal().list().unique_stable().into()
        } else {
            self.into_internal().list().unique().into()
        }
    }
}
